storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java

```diff
@@ -0,0 +1,2250 @@
+        ByteBuffer buf = record.buffer();
+
+        // enforce modify crc to make checksum error
+        buf.put(LegacyRecord.CRC_OFFSET, (byte) 0);
+
+        ByteBuffer buffer = ByteBuffer.allocate(1024);
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, compression,
+                TimestampType.CREATE_TIME, 0L);
+        builder.appendUncheckedWithOffset(0, record);
+
+        MemoryRecords memoryRecords = builder.build();
+        LogValidator logValidator = new LogValidator(memoryRecords,
+                topicPartition,
+                time,
+                type,
+                compression,
+                false,
+                magic,
+                TimestampType.CREATE_TIME,
+                1000L,
+                1000L,
+                RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                AppendOrigin.CLIENT,
+                MetadataVersion.latestTesting()
+        );
+
+
+        assertThrows(CorruptRecordException.class, () -> logValidator.validateMessagesAndAssignOffsets(
+                PrimitiveRef.ofLong(0),
+                metricsRecorder,
+                RequestLocal.withThreadConfinedCaching().bufferSupplier()
+        ));
+
+        assertTrue(metricsRecorder.recordInvalidChecksumsCount > 0);
+    }
+
+    @ParameterizedTest
+    @EnumSource(CompressionType.class)
+    public void testInvalidSequenceV0(CompressionType type) {
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V0, Compression.gzip().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V0, Compression.lz4().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V0, Compression.snappy().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V0, Compression.zstd().build(), type);
+    }
+
+    @ParameterizedTest
+    @EnumSource(CompressionType.class)
+    public void testInvalidSequenceV1(CompressionType type) {
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V1, Compression.gzip().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V1, Compression.lz4().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V1, Compression.snappy().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V1, Compression.zstd().build(), type);
+    }
+
+    @ParameterizedTest
+    @EnumSource(CompressionType.class)
+    public void testInvalidSequenceV2(CompressionType type) {
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V2, Compression.gzip().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V2, Compression.lz4().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V2, Compression.snappy().build(), type);
+        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V2, Compression.zstd().build(), type);
+    }
+
+
+    private void checkInvalidSequence(byte magic, Compression compression, CompressionType type) {
```

Comment:
Could you use `MethodSource`? for example:
```java
    private static Stream<Arguments> testInvalidSequence() {
        return Stream.of(RecordBatch.MAGIC_VALUE_V0, RecordBatch.MAGIC_VALUE_V1, RecordBatch.MAGIC_VALUE_V2)
                .flatMap(code -> Arrays.stream(CompressionType.values()).flatMap(source ->
                        Arrays.stream(CompressionType.values()).map(target -> Arguments.of(code, source, target))));
    }

    @ParameterizedTest
    @MethodSource("testInvalidSequence")
    void testInvalidSequence(byte magic, CompressionType source, CompressionType target) {
```

Reply:
Done.